### PR TITLE
I want to browse and download credit memo

### PR DESCRIPTION
**Pull request by bartoszpietrzak1994**
screen shot 2018-08-07 at 12 25 36

### CODE REVIEW

File: features/downloading_credit_memo_on_customer_order_view.feature [hunk 1]
```diff
@@ -0,0 +1,25 @@
+@customer_browsing_credit_memos
+Feature: Downloading credit memo on customer order view
+    In order to have access to all credit memos related to order refunds
```
Comment (Zales0123) [inline features/downloading_credit_memo_on_customer_order_view.feature]: `In order to save generated credit memos for future usage`?

File: features/downloading_credit_memo_on_customer_order_view.feature [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "Angel T-Shirt"
+        And the store ships everywhere for free
+        And the store allows paying with "Cash on Delivery"
+        And I am a logged in customer
+        And I placed an order "#00000666"
+        And I bought a single "Angel T-Shirt"
+        And I addressed it to "Lucifer Morningstar", "Seaside Fwy", "90802" "Los Angeles" in the "United States"
```
Comment (Zales0123) [inline features/downloading_credit_memo_on_customer_order_view.feature]: `with identical billing address` :)

File: tests/Behat/Page/Shop/Order/ShowPage.php [hunk 3]
```diff
@@ -12,4 +13,38 @@ public function countCreditMemos(): int
...
         return count($this->getDocument()->findAll('css', '#credit-memos tbody tr'));
     }
+
+    public function downloadFirstCreditMemo(): void
+    {
+        $creditMemo = $this->getFirstCreditMemo();
+        $creditMemo->clickLink('Download');
+    }
+
+    public function isPdfFileDownloaded(): bool
```
Comment (Zales0123) [inline tests/Behat/Page/Shop/Order/ShowPage.php]: It should not be implemented here, we have `Tests\Sylius\RefundPlugin\Behat\Context\Element\PdfDownloadElement` (which in fact should be moved to `Tests\Sylius\RefundPlugin\Behat\Element\PdfDownloadElement`)

File: tests/Behat/Page/Shop/Order/ShowPage.php [hunk 4]
```diff
@@ -12,4 +13,38 @@ public function countCreditMemos(): int
     {
         return count($this->getDocument()->findAll('css', '#credit-memos tbody tr'));
     }
+
+    public function downloadFirstCreditMemo(): void
```
Comment (Zales0123) [inline tests/Behat/Page/Shop/Order/ShowPage.php]: What about something more flexible like `downloadCreditMemo(int $index): void`?

File: src/Resources/config/shop_routing.yml [hunk 5]
```diff
@@ -8,3 +8,9 @@ sylius_refund_plugin_shop_order_credit_memos_partial:
             repository:
                 method: findByOrderNumber
                 arguments: $orderNumber
+
+sylius_refund_shop_credit_memo_download:
+    path: /credit-memos/{id}/download
+    methods: [GET]
+    defaults:
+        _controller: Sylius\RefundPlugin\Action\DownloadCreditMemoAction
```
Comment (Zales0123) [inline src/Resources/config/shop_routing.yml]: I think this is a possible security issue - we're using the same action for Admin panel and Customer account, therefore it's possible for a Customer to access any credit memo in the system :( We should create a new action that would check the possibility to download credit memo... There is not so much logic to duplicate :)

File: src/Action/Shop/DownloadCreditMemoAction.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        /** @var CustomerInterface $orderCustomer */
+        $orderCustomer = $order->getCustomer();
+
+        /** @var CustomerInterface $customer */
+        $customer = $this->customerContext->getCustomer();
+
+        if ($orderCustomer->getId() !== $customer->getId()) {
+            return new Response(Response::HTTP_UNAUTHORIZED);
+        }
```
Comment (Zales0123) [inline src/Action/Shop/DownloadCreditMemoAction.php]: This whole logic should be extracted to separate service, that would throw an exception if a customer is not permitted to perform an action

File: src/Action/Shop/DownloadCreditMemoAction.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+        $creditMemoPdfFile = $this->creditMemoPdfFileGenerator->generate($id);
+
+        $response = new Response($creditMemoPdfFile->content(), Response::HTTP_OK, ['Content-Type' => 'application/pdf']);
+        $response->headers->add([
+            'Content-Disposition' => $response->headers->makeDisposition('attachment', $creditMemoPdfFile->filename()),
+        ]);
+
+        return $response;
```
Comment (Zales0123) [inline src/Action/Shop/DownloadCreditMemoAction.php]: I think we can even extract this one to something like `CreditMemoFileResponseBuilder` as it's used in two actions :)

File: src/Checker/CreditMemoCustomerRelationChecker.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->findOneByNumber($creditMemo->getOrderNumber());
+
+        /** @var CustomerInterface $orderCustomer */
+        $orderCustomer = $order->getCustomer();
+
+        /** @var CustomerInterface $customer */
+        $customer = $this->customerContext->getCustomer();
+
+        Assert::eq($orderCustomer->getId(), $customer->getId());
```
Comment (Zales0123) [inline src/Checker/CreditMemoCustomerRelationChecker.php]: We can think about some more sophisticated exception like `CreditMemoNotAccessible` or sth similar 😄

File: src/Checker/CreditMemoCustomerRelationChecker.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    public function check(string $creditMemoId): void
+    {
+        /** @var CreditMemoInterface $creditMemo */
+        $creditMemo = $this->creditMemoRepository->find($creditMemoId);
+
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->findOneByNumber($creditMemo->getOrderNumber());
+
+        /** @var CustomerInterface $orderCustomer */
```
[the remark this answers is not in the document]
Comment (bartoszpietrzak1994) [reply src/Checker/CreditMemoCustomerRelationChecker.php]: Unfortunately, static code analysis does not pass without these type hints.

**Pull request merged**